### DISCUSSION TITLE
Override dependency

### DISCUSSION THREAD

**Discussion by marijanp** [ipetkov/crane#472] [Q&A]
Hello, I'm packaging a crate that is available on crates.io but not in nixpkgs using crane. I'm using crane's `buildPackage` and nixpkgs `fetchCrate` to fetch the sources. One of the dependencies of the package I'm trying to package downloads a bunch of files in its `build.rs`. I was able to figure out what these files are and now I would like to place them where they belong. Is there a way to override the dependency of the package somehow? Ideally I would add a `postUnpack` hook.

The dependency that fetches files from the internet is part of the workspace of the package I'm trying to build.

Thanks for your help in advance

**Comment by ipetkov** [comment]
It will largely depend on what that build script is doing and whether you can turn off the fetch part (assuming you can satisfy it's requirements by making the files available with a `postUnpack` script or equivalent).

If the build script unconditionally tries to touch the network, you'll need to either update it upstream, or have the derivation apply a patch before continuing with the build. 

Hope this helps!

**Reply by marijanp** [reply]
Studying the `build.rs` it reads the packages `src` directory to check whether the file is present. This is what makes me struggle. The source directory after the `unpackPhase` does not look at all like the actual github sources and I'm not able to inject the file where the build script is looking at it. The `build.rs` reads a `src/file.zip`, relative to the packages root directory.

**Reply by marijanp** [reply]
I also tried to replace `fetchCrate` with `fetchFromGitHub` but then I run into toml parsing issues, and there are more problems with the repo like a missing `Cargo.lock`

**Reply by ipetkov** [reply]
> Studying the build.rs it reads the packages src directory to check whether the file is present. This is what makes me struggle. The source directory after the unpackPhase does not look at all like the actual github sources and I'm not able to inject the file where the build script is looking at it. The build.rs reads a src/file.zip, relative to the packages root directory.

It's worth noting that a build script's current working directory [is the source directory of the build script’s package](https://doc.rust-lang.org/cargo/reference/build-scripts.html#inputs-to-the-build-script) and not the source root that the derivation starts out with

---

> I also tried to replace `fetchCrate` with `fetchFromGitHub` but then I run into toml parsing issues, and there are more problems with the repo like a missing `Cargo.lock`

There's some instructions for handling this documented here which might be helpful: https://crane.dev/faq/no-cargo-lock.html

**Reply by marijanp** [reply]
> It's worth noting that a build script's current working directory [is the source directory of the build script’s package](https://doc.rust-lang.org/cargo/reference/build-scripts.html#inputs-to-the-build-script) and not the source root that the derivation starts out with

Is there a way to "patch" the build scripts working directory using crane?

**Reply by ipetkov** [reply]
Yes, all arguments are eventually lowered to `mkDerivation`, meaning you can set `patches` and the `patchPhase` will apply them to the source: https://nixos.org/manual/nixpkgs/stable/#ssec-patch-phase

Alternatively you can define a custom phase to patch things more "manually" (e.g. with a shell or sed script) if that works out better

**Reply by marijanp** [reply]
Ok I might need to make this clearer: I'm familiar with all the concepts in order to achieve patching.
The problem is: after the `unpackPhase` (using `fetchCrate` to obtain the sources) the source directory does not resemble the source directory encounter on the project's GitHub. Instead I'm left with:
```
|-- Cargo.lock
|-- Cargo.toml
`-- src
    |-- bin
    |   `-- crane-dummy-risc0-r0vm
    |       `-- main.rs
    `-- lib.rs
```
If I instead use `fetchFromGithub` crane misinterprets some templates which are part of that repository and when `fromToml` is called I get errors thrown by crane.

**Reply by ipetkov** [reply]
>  (using fetchCrate to obtain the sources) the source directory does not resemble the source directory encounter on the project's GitHub

`fetchCrate` pulls what's published to crates.io; cargo will normally only publish a crate's directory (modulo any explicit `include`/`exclude` definitions). If the crate is within a subdirectory it will likely not match what the project's repo looks like

Also you might be getting errors in trying to patch the dummfied sources during the `buildDepsOnly` derivation. If you are using `buildPackage` directly, consider setting something like:

```nix
let
  commonArgs = {
    # everything except for patching logic
  };
in
craneLib.buildPackage (commonArgs // {
  cargoArtifacts = craneLib.buildDepsOnly commonArgs;
  # patching logic goes here
});
```
